### PR TITLE
Update spaceship to properly handle devices after recent Dev Portal change

### PR DESCRIPTION
**Pull request by KrauseFx**
Related https://github.com/fastlane/fastlane/issues/6137#issuecomment-247591758
https://twitter.com/KrauseFx/status/776887607456178177

### CODE REVIEW

File: spaceship/lib/spaceship/portal/provisioning_profile.rb [hunk 1]
```diff
@@ -164,12 +174,30 @@ def factory(attrs)
...
+          # see https://github.com/fastlane/fastlane/issues/6137 for more information
+          # That's why we set it here
+
+          # Parse all the devices from the details request
+          obj.devices = (profile_details["devices"] || []).collect do |device|
+            Device.set_client(client).factory(device)
+          end
+
+          # Parse all the certificates from the details request
+          obj.certificates = (profile_details["certificates"] || []).collect do |cert|
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/portal/provisioning_profile.rb]: Nope, not preferring this

File: spaceship/lib/spaceship/portal/provisioning_profile.rb [hunk 2]
```diff
@@ -148,7 +153,12 @@ def type
         # This is used to create a new object based on the server response.
         def factory(attrs)
           # Ad Hoc Profiles look exactly like App Store profiles, but usually include devices
-          attrs['distributionMethod'] = 'adhoc' if attrs['distributionMethod'] == 'store' && attrs['devices'].size > 0
+          profile_details = client.provisioning_profile_details(provisioning_profile_id: attrs["provisioningProfileId"])
+
+          # The only difference between App Store and Ad Hoc provisioniong profile is the devices that are attached
+          if attrs['distributionMethod'] == 'store' && (profile_details['devices'] || []).size > 0
```
Comment (mfurtak) [inline spaceship/lib/spaceship/portal/provisioning_profile.rb]: Personally, @houndci-bot, I would say  `!(profile_details['devices'] || []).empty?`

File: spaceship/lib/spaceship/portal/provisioning_profile.rb (hunk 1 above)
Comment (mfurtak) [inline spaceship/lib/spaceship/portal/provisioning_profile.rb]: Shall we disable this warning?

**Review by mfurtak — APPROVED** [review]
Looks good! 👍

File: spaceship/lib/spaceship/portal/provisioning_profile.rb [hunk 3]
```diff
@@ -164,12 +174,37 @@ def factory(attrs)
...
+          # That's why we set it here
+
+          # The reason why we set access attrs["devices"], then profile_details["devices"] is just
+          # that we don't have to re-create all the stubs for the local unit tests
+          # Both contain the same format, but we don't want to update all the JSON and XML files
+          # and all the tests with new IDs, so we're just accessing the data here
+
+          # Parse all the devices from the details request
+          profile_details["devices"] = attrs["devices"] unless (attrs["devices"] || []).empty?
+          obj.devices = (profile_details["devices"] || []).collect do |device|
```
Comment (KrauseFx) [inline spaceship/lib/spaceship/portal/provisioning_profile.rb]: @mfurtak yes, do you know how?

Comment (mfurtak) [inline spaceship/lib/spaceship/portal/provisioning_profile.rb]: I can look into it for another PR 👍

File: spaceship/spec/provisioning_profile_spec.rb [hunk 4]
```diff
@@ -187,14 +179,14 @@
...
-        devices = ["RK3285QATH", "E687498679", "5YTNZ5A9RV", "VCD3RH54BK", "VA3Z744A8R", "T5VFWSCC2Z", "GD25LDGN99", "XJXGVS46MW", "L4378H292Z", "9T5RA84V77", "S4227Y42V5", "LEL449RZER", "WXQ7V239BE"]
+        devices = ["FVRY7XH22J", "4ZE252U553"]
         expect(client).to receive(:repair_provisioning_profile!).with('475ESRP5F3', 'net.sunapps.7 Development', 'limited', '572XTN75U2', [cert_id], devices, mac: false).and_return({})
         profile.repair!
       end
     end
   end
 
-  describe "#update!" do
+  describe "#update!", nower: true do
```
Comment (mfurtak) [inline spaceship/spec/provisioning_profile_spec.rb]: leftover decorator?

**Review by mfurtak — APPROVED** [review]
Approved, though check out the leftover test decorator 👍

File: spaceship/spec/provisioning_profile_spec.rb (hunk 4 above)
Comment (KrauseFx) [inline spaceship/spec/provisioning_profile_spec.rb]: Yes, nice catch

Comment (KrauseFx) [inline spaceship/spec/provisioning_profile_spec.rb]: At least I removed the `yolo: true` 👍

**Pull request merged**